### PR TITLE
ENT-4222 : replace REST api calls to LMS for certificate and grade data fetch

### PR DESCRIPTION
**Pull request by binodpant**
Replace api calls with utils/database calls to fetch grade and cert info

### Not including (see comments below) in this PR:

These two remaining api calls should clear out all LMS api calls from IC code base (leaving just api calls to LMS-external services):
- [ ] Call lms util instead of REST api to get assessment grades (current call is `grades_api.get_course_assessment_grades`)
- [ ] Call lms util instead of REST Api for `get_course_details` , this one uses the `CourseApiClient` whose usage can also be removed from this codebase

Reason:  I want to test this outgoing PR in staging with at least one LMS first then will add a new PR to handle this piece to avoid making this PR too large to test.

### Included work:

- [x] Gather up calls from integrated_channels exporter to lms into integrated_channels.lms_utils file
- [x] Gather up calls from integrated_channels exporter  to catalog service into integrated_channels.catalog_service_utils file
- [x] Call lms util instead of REST api to get certificate for user 
- [x] Calls lms util instead of RET api to implement `_collect_grades_data`
- [x] Refactor code for better readability and reduced complexity
- [x] Refactor out some logging calls to improve readability and better prep for future logging enhancements
- [x] Removes the already unused (`self.enterprise_api`) property that refers to `EnterpriseApiClient` in Exporter 

**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated
- [ ] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.
- [ ] [Version](https://github.com/edx/edx-enterprise/blob/master/enterprise/__init__.py) bumped
- [ ] [Changelog](https://github.com/edx/edx-enterprise/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/edx/edx-enterprise/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [GitHub Actions](https://github.com/edx/edx-enterprise/actions), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-platform](https://github.com/edx/edx-platform) to upgrade dependencies (including edx-enterprise)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

### CODE REVIEW

**Review by alex-sheehan-edx — COMMENTED** [review]
A couple comments, I also think there are some tests that need to be fixed, but overall love the cleanup!

File: integrated_channels/integrated_channel/exporters/learner_data.py [hunk 1]
```diff
@@ -253,27 +254,12 @@ def export(self, **kwargs):  # pylint: disable=R0915
             channel_name=channel_name,
             enterprise_customer_identifier=self.enterprise_customer.name
         )
-        enrollment_queryset = EnterpriseCourseEnrollment.objects.select_related(
```
[the remark this answers is not in the document]
Comment (binodpant) [reply integrated_channels/integrated_channel/exporters/learner_data.py:256]: it was not removed, just refactored into a separate method for better readability in this method, the `create_enrollment_queryset` was extracted from the removed blocks of code here

[the remark this answers is not in the document]
Comment (binodpant) [reply integrated_channels/integrated_channel/exporters/learner_data.py:256]: another related reason was I wanted to simplify what happens in the for loop that follows, so now that loop runs on a pre filtered set, rather than doing filtering within the loop

File: tests/test_integrated_channels/test_integrated_channel/test_exporters/test_learner_data.py [hunk 2]
```diff
@@ -675,7 +673,7 @@ def test_learner_exporter_with_skip_transmitted(
         # Check that LMS enrollment populated as part of model used in audit check:
         expected_result = mock_course_enrollment_class.objects.get.return_value
         self.assertEqual(expected_result, enterprise_course_enrollment.course_enrollment)
-        assert mock_course_enrollment_class.objects.get.call_count == 2
```
Comment (binodpant) [inline tests/test_integrated_channels/test_integrated_channel/test_exporters/test_learner_data.py:678]: I am not 100% sure why this mock call count went from 2 to 1. but I did debug through the code while running the test and verified that the queryset is correct and the rest of the stuff works as expected. Ideas?

**Comment by binodpant** [comment]
@alex-sheehan-edx  @edx-abolger  I am done replacing api calls for two of the fetches in `learner_data.py`, but I think the last one `_collect_assessment_grades_data`  is going to be a fair amount of work due to the amount of logic contained in the endpoint it uses which is here https://github.com/edx/edx-platform/blob/96f0915b0fbde221dadb0b37fba6f186894260b4/lms/djangoapps/grades/rest_api/v1/gradebook_views.py#L511. 


I suggest we do that separately in a new ticket, I have already been working on this ticket a few days, think there’s good improvement here already

File: integrated_channels/integrated_channel/exporters/learner_data.py [hunk 3]
```diff
@@ -439,6 +426,60 @@ def export(self, **kwargs):  # pylint: disable=R0915
...
+    ):
+        """
+        Given an enrollment_queryset, returns only enrollments that are not already transmitted
+        """
+        included_enrollments = set()
+        for enterprise_enrollment in enrollment_queryset:
+            if transmission_audit and \
+                    is_already_transmitted(transmission_audit, enterprise_enrollment.id, grade):
+                # We've already sent a completion status for this enrollment
+                generate_formatted_log(
```
Comment (binodpant) [inline integrated_channels/integrated_channel/exporters/learner_data.py:444]: for now we can keep this log per item, I want to revisit this as part of the logging cleanup

File: integrated_channels/catalog_service_utils.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
+"""
+A utility collection for calls from integrated_channels to Catalog service
+"""
+from enterprise.api_client.discovery import get_course_catalog_api_service_client
+
+
+def get_course_id_for_enrollment(enterprise_enrollment):
+    """
+    Fetch course_id for a given enterprise enrollment
```
Comment (edx-abolger) [inline integrated_channels/catalog_service_utils.py:9]: For both functions please add a docstring about what happens if one is not found (since this is a remote client): specific exception, returns none?

File: integrated_channels/integrated_channel/exporters/learner_data.py [hunk 5]
```diff
@@ -287,21 +273,19 @@ def export(self, **kwargs):  # pylint: disable=R0915
...
-                generate_formatted_log(
-                    'Skipping export of previously sent enterprise enrollment. '
-                    'EnterpriseEnrollment: {enterprise_enrollment_id}'.format(
-                        enterprise_enrollment_id=enterprise_enrollment.id
-                    ),
-                    channel_name=channel_name,
-                    enterprise_customer_identifier=self.enterprise_customer.name
-                )
-                continue
+        if TransmissionAudit and skip_transmitted:
```
Comment (edx-abolger) [inline integrated_channels/integrated_channel/exporters/learner_data.py:276]: This refactor seems logical, does it pass any existing tests (are there existing tests for the export() function?)

File: integrated_channels/integrated_channel/exporters/learner_data.py [hunk 6]
```diff
@@ -287,21 +273,19 @@ def export(self, **kwargs):  # pylint: disable=R0915
...
+            enrollments_to_transmit = self._filter_out_pre_transmitted_enrollments(
+                enrollment_queryset,
+                channel_name,
+                grade,
+                TransmissionAudit
+            )
+        else:
+            enrollments_to_transmit = enrollment_queryset
```
Comment (edx-abolger) [inline integrated_channels/integrated_channel/exporters/learner_data.py:284]: The one thing that I don't like about this refactor is that your new _filter_out_pre_transmitted_enrollments function actually _resolves_ the query set (ie will make a database call) because it iterates over the results and this means that we could get slightly different behaviors at line 286, since if the code path goes through 284 it will have a different query.  That's one reason this was likely all kept together in the first place: your refactor will result in a hidden side effect that might make debugging later harder than needed.  

Basically, can you please change your _filter_out_pre_transmitted_enrollments to accept a _resolved_ queryset, so instead of passing it enrollment_queryset you pass it the result, like 'all_enrollments = enrollment_queryset.list()'  and then pass in 'all_enrollments'. And then you can change line 284 to also use 'all_enrollments' ?

Comment (edx-abolger) [reply integrated_channels/integrated_channel/exporters/learner_data.py:284]: Further reading in this area: https://docs.djangoproject.com/en/3.1/ref/models/querysets/

File: integrated_channels/integrated_channel/exporters/learner_data.py [hunk 7]
```diff
@@ -375,12 +359,15 @@ def export(self, **kwargs):  # pylint: disable=R0915
...
                         course_id=course_id,
-                        user_id=enterprise_enrollment.enterprise_customer_user.user_id,
+                        user_id=enterprise_user_id,
                         enterprise=enterprise_enrollment.enterprise_customer_user.enterprise_customer.slug
                     ),
                     channel_name=channel_name,
                     enterprise_customer_identifier=self.enterprise_customer.name
                 )
+
+            # todo: we need to refactor so single learner is not so mixed in with bulk case
```
Comment (edx-abolger) [inline integrated_channels/integrated_channel/exporters/learner_data.py:369]: Please do not leave TODOs in code: file a ticket and leave the ticket here if necessary.

File: integrated_channels/integrated_channel/exporters/learner_data.py [hunk 8]
```diff
@@ -439,6 +426,60 @@ def export(self, **kwargs):  # pylint: disable=R0915
...
+            enterprise_customer_user__enterprise_customer=self.enterprise_customer,
+            enterprise_customer_user__active=True,
+        )
+        if learner_to_transmit and course_run_id:
+            enrollment_queryset = enrollment_queryset.filter(
+                course_id=course_run_id,
+                enterprise_customer_user__user_id=learner_to_transmit.id,
+            )
+            generate_formatted_log(
+                'Exporting single learner. Course: {course_run}, User: {user_id}'.format(
```
Comment (edx-abolger) [inline integrated_channels/integrated_channel/exporters/learner_data.py:473]: If this is the lms_user_id can you label it as the LMS User Id so that looking at logs later no one has to go to the code to look up what the number means?

File: integrated_channels/integrated_channel/exporters/learner_data.py [hunk 9]
```diff
@@ -516,39 +557,64 @@ def _collect_certificate_data(self, enterprise_enrollment):
...
+        grade = self.grade_incomplete
+        is_passing = False
+        percent_grade = None
 
-            # For consistency with _collect_grades_data, we only care about Pass/Fail grades. This could change.
-            is_passing = certificate.get('is_passing')
-            percent_grade = certificate.get('grade')
-            grade = self.grade_passing if is_passing else self.grade_failing
+        try:
+            certificate = get_course_certificate(course_id, username)
```
Comment (edx-abolger) [inline integrated_channels/integrated_channel/exporters/learner_data.py:570]: Why do we get course certificate by Username here? We should always be getting these items by lms_user_id.

File: integrated_channels/integrated_channel/exporters/learner_data.py [hunk 10]
```diff
@@ -516,39 +557,64 @@ def _collect_certificate_data(self, enterprise_enrollment):
...
+            return completed_date, grade, is_passing, percent_grade
+
+        completed_date = certificate.get('created_date')
+        if completed_date:
+            completed_date = parse_datetime(completed_date)
+        else:
+            completed_date = timezone.now()
 
-        except HttpNotFoundError:
-            LOGGER.error('[Integrated Channel] Certificate data not found.'
```
Comment (edx-abolger) [inline integrated_channels/integrated_channel/exporters/learner_data.py:539]: Thank you for removing the PII from this log!

File: integrated_channels/integrated_channel/exporters/learner_data.py [hunk 11]
```diff
@@ -516,39 +557,64 @@ def _collect_certificate_data(self, enterprise_enrollment):
...
+        # For consistency with _collect_grades_data, we only care about Pass/Fail grades. This could change.
+        is_passing = certificate.get('is_passing')
+        percent_grade = certificate.get('grade')
+        grade = self.grade_passing if is_passing else self.grade_failing
 
         return completed_date, grade, is_passing, percent_grade
 
+    def _log_cert_not_found(self, course_id, enterprise_enrollment, user_id):
+        """
+        Standardized logging for no certificate found (refactor candidate)
```
Comment (edx-abolger) [inline integrated_channels/integrated_channel/exporters/learner_data.py:594]: Again, please file tickets rather than leaving 'todo' like comment in the code.

File: integrated_channels/integrated_channel/exporters/learner_data.py [hunk 12]
```diff
@@ -516,39 +557,64 @@ def _collect_certificate_data(self, enterprise_enrollment):
...
-            is_passing = False
-            percent_grade = None
+        # For consistency with _collect_grades_data, we only care about Pass/Fail grades. This could change.
+        is_passing = certificate.get('is_passing')
+        percent_grade = certificate.get('grade')
+        grade = self.grade_passing if is_passing else self.grade_failing
 
         return completed_date, grade, is_passing, percent_grade
 
+    def _log_cert_not_found(self, course_id, enterprise_enrollment, user_id):
```
Comment (edx-abolger) [inline integrated_channels/integrated_channel/exporters/learner_data.py:592]: Nit: can we change 'user_id' to 'lms_user_id' everywhere?

File: integrated_channels/integrated_channel/exporters/learner_data.py [hunk 13]
```diff
@@ -516,39 +557,64 @@ def _collect_certificate_data(self, enterprise_enrollment):
...
+        """
+        Standardized logging for no certificate found (refactor candidate)
+        """
+        LOGGER.error('[Integrated Channel] Certificate not found for user'
+                     ' Course: {course_id}, EnterpriseEnrollment: {enterprise_enrollment}, '
+                     ' Learner id {user_id}'
```
Comment (edx-abolger) [inline integrated_channels/integrated_channel/exporters/learner_data.py:598]: Nit: can we label as Learner LMS User Id in the log so it's easy to tell where to go look this up?

File: integrated_channels/integrated_channel/exporters/learner_data.py [hunk 14]
```diff
@@ -608,39 +674,24 @@ def _collect_grades_data(self, enterprise_enrollment, course_details, is_audit_e
...
-                    # This means the user has an enterprise enrollment record but is not enrolled in the course yet
-                    LOGGER.info(
-                        '[Integrated Channel] User is not enrolled in the course.'
-                        ' Course: {course_id}, EnterpriseEnrollment: {enterprise_enrollment},'
-                        ' Username: {username}'.format(
-                            course_id=course_id,
-                            username=username,
-                            enterprise_enrollment=enterprise_enrollment.pk))
-                    return None, None, None, None
+        grades_data = get_single_user_grade(course_id, username)
```
Comment (edx-abolger) [inline integrated_channels/integrated_channel/exporters/learner_data.py:687]: Again, can we  use user_id to get this information instead of username?

File: integrated_channels/lms_utils.py [hunk 15]
```diff
@@ -0,0 +1,55 @@
+"""
```
Comment (edx-abolger) [inline integrated_channels/lms_utils.py:1]: This new module needs tests.

File: integrated_channels/integrated_channel/exporters/learner_data.py [hunk 16]
```diff
@@ -608,39 +674,24 @@ def _collect_grades_data(self, enterprise_enrollment, course_details, is_audit_e
...
 
+        if grades_data is None:
             LOGGER.error('[Integrated Channel] Grades data not found.'
                          ' Course: {course_id}, EnterpriseEnrollment: {enterprise_enrollment},'
-                         ' Username: {username}'.format(
+                         ' Userid: {user_id}'.format(
```
Comment (edx-abolger) [inline integrated_channels/integrated_channel/exporters/learner_data.py:692]: +1 for PII removal, thank you!

**Review by edx-abolger — CHANGES_REQUESTED** [review]
First, great job removing PII from logs!  

Changes requested: 
1.  main issue is lack of small unit tests for new modules.  You should be able to make the tests look similar to existing tests that make sure the right database model value comes back. 
2.  a few log message update requests
3. one small change to a refactor you did to not pass in a queryset as a parameter

File: integrated_channels/integrated_channel/exporters/learner_data.py [hunk 17]
```diff
@@ -9,16 +9,18 @@
...
+from opaque_keys import InvalidKeyError
 from slumber.exceptions import HttpNotFoundError
 
 from django.apps import apps
 from django.utils import timezone
 from django.utils.dateparse import parse_datetime
 
 from consent.models import DataSharingConsent
-from enterprise.api_client.lms import CertificatesApiClient, CourseApiClient, GradesApiClient
+from enterprise.api_client.lms import CourseApiClient, GradesApiClient
```
[the remark this answers is not in the document]
Comment (binodpant) [reply integrated_channels/integrated_channel/exporters/learner_data.py:20]: exactly yes!

File: integrated_channels/integrated_channel/exporters/learner_data.py (hunk 5 above)
Comment (binodpant) [reply integrated_channels/integrated_channel/exporters/learner_data.py:276]: yup it passes existing tests (in fact I had to edit existing tests to ensure they passes as they no longer can use the mocks for the api clients

File: integrated_channels/integrated_channel/exporters/learner_data.py (hunk 6 above)
Comment (binodpant) [reply integrated_channels/integrated_channel/exporters/learner_data.py:284]: got it had to read it twice but yes I had a lingering thought in this area too, I think it's best to avoid the discrepancy.. will take a look thanks

File: integrated_channels/integrated_channel/exporters/learner_data.py (hunk 14 above)
Comment (binodpant) [reply integrated_channels/integrated_channel/exporters/learner_data.py:687]: ah not sure thatt will work given the api call I am using is expecting a username it appears

```
CourseGradeFactory().read(grade_user, course_key=course_key)
```

Comment (edx-abolger) [reply integrated_channels/integrated_channel/exporters/learner_data.py:687]: That's a "user" as in the object "Auth.user", not a string. I think the user id should be used to get you a user and username shouldn't work anyway, let me look into that for a second.

Comment (edx-abolger) [reply integrated_channels/integrated_channel/exporters/learner_data.py:687]: Alright so the original call you replaced used the username for who knows what reason. But you don't have to do that, and anyway the grade_user you are looking at there is definitely a user object, not a username string. So you can use django to get the correct user by id and pass it in, and just remove the username usage here completely.

Comment (binodpant) [reply integrated_channels/integrated_channel/exporters/learner_data.py]: got it, now passing the user object instead!

**Review by edx-abolger — APPROVED** [review]
Would still like tickets filed for todos added to comments, but changes are approved.
